enclave/core/sgx/init.c

```diff
@@ -52,6 +52,14 @@ static int _is_eeid_base_image(
            properties->header.size_settings.num_tcs == 1;
 }
 
+static int _is_within_eeid_enclave(const uint8_t* ptr)
```

Comment:
OE SDK already has a _oe_is_within_enclave(const void* p, size_t n)_ API. No need to add a EEID-specific version. _oe_is_within_enclave(...)_ gets the enclave size from oe_enclave_properties_sgx.image_info.enclave_size, which should have been set to EEID_SECS_SIZE in EEID case.

Reply:
No, that's not the case. `oe_enclave_properties_sgx.image_info.enclave_size` holds the base image enclave size (i.e. heap/stack == 0). Note that there are two calls to `_calculate_enclave_size` and only the result of the first one is patched into the image. For EEID, we patch that same base-image value into that variable, because otherwise we would change the base image hash. I couldn't find a way to get the SECS data directly, but conveniently, we always set it to the maximum anyways.

Reply:
In create.c:

    /* Calculate the size of this enclave in memory */
    OE_CHECK(_calculate_enclave_size(
        image_size, &props, &enclave_end, &enclave_size));

    /* Perform the ECREATE operation */
    OE_CHECK(oe_sgx_create_enclave(
        context, enclave_size, enclave_end, &enclave_addr));

    /* Save the enclave base address, size, and text address */
    enclave->addr = enclave_addr;
    enclave->size = enclave_size;
    enclave->text = enclave_addr + oeimage.text_rva;

    /* Patch image */
    OE_CHECK(oeimage.patch(&oeimage, enclave_size));

I thought _calculate_enclave_size() set enclave_size to EEID_SECS_SIZE (for EEID case) and the value is used byoeimage.patch to set oe_enclave_properties_sgx.image_info.enclave_size.

Anyway, if oe_enclave_properties_sgx.image_info.enclave_size indeed holds the base image enclave size (heap/stack=0), it needs to be fixed. The enclave code also uses it to check whether an output buffer that is supposed to be outside of enclave is indeed outside of enclave. If that check is not reliable, the enclave code can unintentionally overwrite its own heap or stack.

Reply:
You may be right, I have to read the code again as this change a few times.

Reply:
You're right, it is in fact patched into the base image too now, so it should hold the right size during enclave entry.

Reply:
Done!